source/simulator/assemblers/advection.cc

```diff
@@ -503,6 +504,110 @@ namespace aspect
+      Assert (scratch.face_phi_field.size() == advection_dofs_per_cell, ExcInternalError());
+
+      const unsigned int solution_component = advection_field.component_index(introspection);
+      const FEValuesExtractors::Scalar solution_field = advection_field.scalar_extractor(introspection);
+
+      if (this->get_fixed_convective_heating_boundary_indicators().find(face->boundary_id())
+          != this->get_fixed_convective_heating_boundary_indicators().end())
+        {
+          // We are in the case of a Robin temperature boundary.
+          // Impose the Robin value weakly using both a LHS and a RHS term.
+
+          const std::vector<Tensor<1,dim>> heat_flux
+            = this->get_boundary_convective_heating_manager().heat_flux(
+                face->boundary_id(),
+                scratch.face_material_model_inputs,
+                scratch.face_material_model_outputs,
+                scratch.face_finite_element_values->get_normal_vectors());
+
+          const std::vector<double> heat_transfer_coefficients
+            = this->get_boundary_convective_heating_manager().heat_transfer_coefficient(
+                face->boundary_id(),
+                scratch.face_material_model_inputs,
+                scratch.face_material_model_outputs);
+
+          for (unsigned int q=0; q<n_face_q_points; ++q)
+            {
+              // precompute the values of shape functions.
+              // We only need to look up values of shape functions if they
+              // belong to 'our' component. They are zero otherwise anyway.
+              // Note that we later only look at the values that we do set here.
+              for (unsigned int i=0, i_advection=0; i_advection<advection_dofs_per_cell; /*increment at end of loop*/)
+                {
+                  if (fe.system_to_component_index(i).first == solution_component)
+                    {
+                      scratch.face_phi_field[i_advection] = (*scratch.face_finite_element_values)[solution_field].value (i, q);
+                      ++i_advection;
+                    }
+                  ++i;
+                }
+
+              Assert (heat_transfer_coefficients[q] >= 0,
+                      ExcMessage ("The heat transfer coefficient needs to be a "
+                                  "non-negative quantity but instead it is " +
+                                  std::to_string(heat_transfer_coefficients[q]) + "."));
+
+              const double boundary_temperature
+                = this->get_boundary_convective_heating_manager().boundary_temperature(
+                    face->boundary_id(),
+                    scratch.face_finite_element_values->quadrature_point(q));
+
+              for (unsigned int i=0; i<advection_dofs_per_cell; ++i)
+                {
+                  data.local_rhs(i)
+                  += (
+                       -time_step * scratch.face_phi_field[i] *
```

Comment:
I am not sure why that is negative.

Reply:
I struggled with this for a while. The sign depends on how you define the heat flux. I implemented this term to be the same as in the Neumann/Heat flux boundary conditions:

https://github.com/geodynamics/aspect/blob/2d63fc8a51b0471fdf5105282f4e8b8a4f85cce9/source/simulator/assemblers/advection.cc#L492-L498

We define heat flux to be positive if it flows out of the model. So I assume in the weak form I used, it's the opposite.